[FLINK-24038] Add support for single leader election per JobManager process

tillrohrmann opened the pull request:
This PR adds two new high availability services `ZooKeeperMultipleComponentLeaderElectionHaServices` and `KuberntesMultipleComponentLeaderElectionHaServices` that only use a single leader election per process. Both services internally rely on the `DefaultMultipleComponentLeaderElectionService` that uses a single `MultipleComponentLeaderElectionDriver` (responsible for the leader election) and allows to register multiple components. All components either obtain together the leadership or lose it together.

For the `MultipleComponentLeaderElectionDriver`, there are two implementations `ZooKeeperMultipleComponentLeaderElectionDriver` and the `KubernetesMultipleComponentLeaderElectionDriver` that implement the leader election based on ZooKeeper and Kubernetes, respectively.

In order to keep the old ha services implementation for safety purposes (users can switch back by setting `high-availability.use-old-ha-services: true`), I have reused the `DefaultLeaderElectionService` implementation and introduced a `MultipleComponentLeaderElectionDriverAdapter` that internally registers a component at the `MultipleComponentLeaderElectionService`. Once we remove the old HA service implementations I think that we can remove these temporary bridges and simplify the `DefaultLeaderElectionService` implementation to directly work against the `MultipleComponentLeaderElectionService`.

XComp commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesTestFixture.java:
```diff
@@ -0,0 +1,236 @@
+
+    FlinkKubeClient.WatchCallbackHandler<KubernetesConfigMap> getLeaderElectionConfigMapCallback()
+            throws Exception {
+        assertThat(configMapCallbackFutures.size(), is(greaterThanOrEqualTo(1)));
```
I'm also not 100% sure whether we're mixing up test code and assertions here in an unfavorable way? Reading the code I'm not certain why we do the assert here independent of any test method 🤔 I feel like we're assuming some context here that is not necessarily there.

tillrohrmann replied:
We probably are. I am also not super happy with the current state but I also didn't want to completely refactor the existing K8s tests. Hence, I would suggest to create a follow up task for cleaning this up.

XComp commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -56,59 +62,179 @@
+                            assertThat(
+                                            allKnownLeaderInformationEvent
+                                                    .getLeaderInformationWithComponentIds())
+                                    .contains(
```
`containsExactlyInAnyOrder` would be more precise here

XComp commented on flink-core/src/main/java/org/apache/flink/configuration/HighAvailabilityOptions.java:
```diff
@@ -204,6 +204,23 @@
+                    .withDescription(
+                            Description.builder()
+                                    .text(
+                                            "Use this option to disable the new HA service implementations for ZooKeeper and K8s. This is safety hatch in case that the new ha services are buggy.")
```
nit: "[...] This is a safety hatch [...]"

tillrohrmann replied:
Will correct it.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/HighAvailabilityServicesUtils.java:
```diff
@@ -82,6 +80,31 @@ public static HighAvailabilityServices createAvailableOrEmbeddedServices(
         }
     }
 
+    @Nonnull
```
Isn't the `@Nonnull` obsolete here again?

tillrohrmann replied:
Yes. Will remove it.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/HighAvailabilityServicesUtils.java:
```diff
@@ -82,6 +80,31 @@ public static HighAvailabilityServices createAvailableOrEmbeddedServices(
     }
 
+    @Nonnull
+    private static HighAvailabilityServices createZooKeeperHaServices(
```
Why is there a `HighAvailabilityServicesFactory` implementation for k8s but not for ZooKeeper?

tillrohrmann replied:
Because you can only activate the K8s ha services by configuring a factory. This has historical reasons.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionService.java:
```diff
@@ -0,0 +1,285 @@
+            unregisteredLeaderElectionEventHandler =
+                    leaderElectionEventHandlers.remove(componentId);
+
+            if (unregisteredLeaderElectionEventHandler != null) {
```
True. Will add it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/MultipleComponentLeaderElectionService.java:
```diff
@@ -0,0 +1,76 @@
+ * LeaderElectionEventHandlers} that are identified by different names. For each event handler it is
+ * possible to write the corresponding {@link LeaderInformation}.
+ */
+public interface MultipleComponentLeaderElectionService {
```
This is probably a good idea. Will add it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionService.java:
```diff
@@ -0,0 +1,285 @@
+    @GuardedBy("lock")
+    private final Map<String, LeaderElectionEventHandler> leaderElectionEventHandlers;
+
+    private boolean running = true;
```
Good catch. Will change it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionService.java:
```diff
@@ -0,0 +1,285 @@
+            }
+            running = false;
+
+            LOG.info("Closing {}.", this);
```
I think I wanted to put `this.getClass().getSimpleName()` there. Will update it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionService.java:
```diff
@@ -0,0 +1,285 @@
+
+            ExecutorUtils.gracefulShutdown(10L, TimeUnit.SECONDS, leadershipOperationExecutor);
+
+            Exception exception = null;
+            try {
+                multipleComponentLeaderElectionDriver.close();
+            } catch (Exception e) {
+                exception = e;
+            }
+
+            ExceptionUtils.tryRethrowException(exception);
```
No, you are right. I will update it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionService.java:
```diff
@@ -0,0 +1,285 @@
+                fatalErrorHandler,
+                leaderContenderDescription,
+                multipleComponentLeaderElectionDriverFactory,
+                java.util.concurrent.Executors.newSingleThreadExecutor(
```
Indeed. Will update it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/MultipleComponentLeaderElectionDriverFactory.java:
```diff
@@ -0,0 +1,38 @@
+
+    /**
+     * Creates a {@link MultipleComponentLeaderElectionDriver} for the given leader contender
+     * description. Moreover, it registers the given leader election listener with the service.
```
The `MultipleComponentLeaderElectionDriverAdapter` is an implementation of the `LeaderElectionDriver` and not the `MultipleComponentLeaderElectionDriver`. Therefore, I think the comment makes sense. It refers to the fact that the given listener will be registered with the created MCLElectionDriver.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/MultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,87 @@
+         *
+         * @param leaderInformationWithComponentIds leader information with component ids
+         */
+        void notifyAllKnownLeaderInformation(
```
Yes, the different implementations manifest in different interface methods :-(

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionService.java:
```diff
@@ -0,0 +1,285 @@
+                if (leaderInformationByName.containsKey(leaderName)) {
+                    leaderNameLeaderElectionEventHandlerPair
+                            .getValue()
+                            .onLeaderInformationChange(leaderInformationByName.get(leaderName));
```
I think we filter out these cases on a higher level.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,114 @@
+        private BiConsumerWithException<String, LeaderInformation, Exception>
+                publishLeaderInformationConsumer = (ignoredA, ignoredB) -> {};
+        private ThrowingConsumer<String, Exception> deleteLeaderInformationConsumer =
+                (ignored) -> {};
```
True. Will update it.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/TestingMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,114 @@
+    private final ThrowingConsumer<String, Exception> deleteLeaderInformationConsumer;
+    private boolean hasLeadership;
+
+    private Optional<Listener> listener;
```
I think this is because of your inspection settings.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionServiceTest.java:
```diff
@@ -0,0 +1,297 @@
+        }
+    }
+
+    @Nonnull
```
Yes, I will correct it.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionServiceTest.java:
```diff
@@ -0,0 +1,297 @@
+            createDefaultMultiplexingLeaderElectionService(
+                    TestingMultipleComponentLeaderElectionDriver leaderElectionDriver)
+                    throws Exception {
+        final DefaultMultipleComponentLeaderElectionService leaderElectionService =
```
Yes, I will update the code.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionServiceTest.java:
```diff
@@ -0,0 +1,297 @@
+            leaderElectionDriver.grantLeadership();
+
+            final Collection<Component> knownLeaderInformation = createComponents(3);
+            final Collection<Component> unknownLeaderInformation = createComponents(2);
```
I think I wanted to also register these components. Just not send their leader information in the `notifyAllKnownLeaderInformation`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionServiceTest.java:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    @Test
+    public void newlyRegisteredEventHandlersAreInformedAboutLeadership() throws Exception {
```
I am not sure whether this is really needed. I think it is fine to leave this undefined for the time being.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/zookeeper/ZooKeeperMultipleComponentLeaderElectionHaServices.java:
```diff
@@ -0,0 +1,179 @@
+        ExceptionUtils.tryRethrowException(exception);
+    }
+
+    @Override
+    protected void internalCleanupJobData(JobID jobID) throws Exception {
+        super.internalCleanupJobData(jobID);
+    }
```
True. I will remove it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/zookeeper/ZooKeeperMultipleComponentLeaderElectionHaServices.java:
```diff
@@ -0,0 +1,179 @@
+    protected LeaderElectionService createLeaderElectionService(String leaderName) {
+        final MultipleComponentLeaderElectionService multipleComponentLeaderElectionService;
+
+        synchronized (lock) {
```
Yes and probably not. Will change it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/zookeeper/AbstractZooKeeperHaServices.java:
```diff
@@ -0,0 +1,152 @@
+        return ZKPaths.PATH_SEPARATOR.equals(remainingPath);
+    }
+
+    @Nonnull
```
Will remove it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/zookeeper/AbstractZooKeeperHaServices.java:
```diff
@@ -0,0 +1,152 @@
+        return ZKPaths.makePath(path, "");
+    }
+
+    @Nonnull
```
Will remove it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/zookeeper/ZooKeeperMultipleComponentLeaderElectionHaServices.java:
```diff
@@ -0,0 +1,179 @@
+                                new ZooKeeperMultipleComponentLeaderElectionDriverFactory(
+                                        leaderNamespacedCuratorFramework));
+            } catch (Exception e) {
+                throw new FlinkRuntimeException(
```
Yes, the `fatalErrorHandler` is used for asynchronous exception that might occur. When doing synchronous call, I think it is better to directly throw the exception so that the caller can handle it appropriately.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/zookeeper/ZooKeeperMultipleComponentLeaderElectionHaServices.java:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    @Override
+    protected LeaderRetrievalService createLeaderRetrievalService(String leaderPath) {
```
Yes, I think what we would save by moving it to the `AbstractZooKeeperHaServices` does not offset the potential costs that the leader election and retrieval service are not close together which makes it easier to keep them in sync.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java:
```diff
@@ -96,13 +106,17 @@
-    private static final String DISPATCHER_LEADER = "/dispatcher";
+    private static final String DISPATCHER_LEADER = "dispatcher";
 
     private static final String LEADER_NODE = "/leader";
```
True. This is a nice unification. Will apply it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java:
```diff
@@ -96,13 +106,17 @@
-    private static final String REST_SERVER_LEADER = "/rest_server";
+    private static final String REST_SERVER_LEADER = "rest_server";
+
+    private static final String LEADER_LATCH_NODE = "/latch";
```
Yes. Will update it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/zookeeper/ZooKeeperMultipleComponentLeaderElectionHaServices.java:
```diff
@@ -0,0 +1,179 @@
+        super.internalCleanupJobData(jobID);
+    }
+
+    @Override
+    protected String getLeaderPathForResourceManager() {
+        return ZooKeeperUtils.getResourceManagerNode();
+    }
+
+    @Override
+    protected String getLeaderPathForDispatcher() {
+        return ZooKeeperUtils.getDispatcherNode();
+    }
+
+    @Override
+    protected String getLeaderPathForJobManager(JobID jobID) {
+        return jobID.toString();
+    }
+
+    @Override
+    protected String getLeaderPathForRestServer() {
+        return ZooKeeperUtils.getRestServerNode();
+    }
```
I think you are right that there is a difference between the method name and what is returned. In this case I would suggest to change the naming of the methods to `getLeaderNameXYZ` instead of `getLeaderPathXZY`.

tillrohrmann replied:
Arg, the problem is that the different `AbstractHaServices` implementations treat this method differently. Since I don't want to touch the `ZooKeeperHaServices`, I will keep it as is. Once `ZooKeeperHaServices` is removed, we can clean it up.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java:
```diff
@@ -96,13 +106,17 @@
     public static final String HA_STORAGE_COMPLETED_CHECKPOINT = "completedCheckpoint";
 
-    private static final String RESOURCE_MANAGER_LEADER = "/resource_manager";
+    private static final String RESOURCE_MANAGER_LEADER = "resource_manager";
```
This has been addressed in a previous fixup.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java:
```diff
@@ -137,17 +155,37 @@ public static String getCheckpointIdCounterPath() {
         return generateZookeeperPath(LEADER_NODE, suffix);
     }
 
     @Nonnull
```
True, but I haven't touched this code in this PR. Can be a follow-up task.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,250 @@
+            String leaderContenderDescription,
+            MultipleComponentLeaderElectionDriver.Listener leaderElectionListener)
+            throws Exception {
+        this.curatorFramework = curatorFramework;
+        this.leaderContenderDescription = leaderContenderDescription;
+        this.leaderElectionListener = leaderElectionListener;
```
I am a bit torn here. The general Flink guideline says that arguments that are not annotated with `@Nullable` can be assumed to be non-null. Hence, a null check should not strictly be necessary. On the other hand, failing fast in case of a NPE is always a good idea.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,250 @@
+        return leaderInformation;
+    }
+
+    private static class ConnectionInfoNodeSelector implements TreeCacheSelector {
```
Will do.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,250 @@
+        if (running.get() && leaderLatch.hasLeadership() && isConnectionInfoNode(childData)) {
+
+            final String path = childData.getPath();
+            final String[] splits = path.split("/");
```
Agreed for the splitting. The leaderName retrieval is specific to the driver and the used ha services. That's why I would like to keep it here.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,250 @@
+                                case NODE_ADDED:
+                                case NODE_REMOVED:
+                                case NODE_UPDATED:
+                                    if (event.getData() != null) {
```
Yes, I think you are right that this should be a `checkState` for the `event.getData() != null`. Moreover, I think that I am handling the `NODE_REMOVED` event wrongly. In this case, the leader information should be marked as `LeaderInformation.empty` because the event will be created with the old node data. I will correct the code.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,250 @@
+        ZooKeeperUtils.deleteZNode(curatorFramework, ZooKeeperUtils.makeZooKeeperPath(leaderName));
+    }
+
+    private void handleStateChange(ConnectionState newState) {
```
I would like to keep it because the `ZooKeeperLeaderElectionDriver` should ideally be removed with the next release.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -0,0 +1,432 @@
+        return leaderElectionDriver;
+    }
+
+    @Nonnull
```
Will remove it.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -0,0 +1,432 @@
+                Collection<LeaderInformationWithComponentId> leaderInformationWithComponentIds) {}
+    }
+
+    @Nonnull
```
Will remove it.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -0,0 +1,432 @@
+            MultipleComponentLeaderElectionDriver.Listener leaderElectionListener,
+            CuratorFramework curatorFramework)
+            throws Exception {
+        final ZooKeeperMultipleComponentLeaderElectionDriver leaderElectionDriver =
```
True, will remove it.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -0,0 +1,432 @@
+        final Configuration configuration = new Configuration();
+        configuration.set(
+                HighAvailabilityOptions.HA_ZOOKEEPER_QUORUM, zooKeeperExtension.getConnectString());
+        final CuratorFrameworkWithUnhandledErrorListener curatorFramework =
```
Will correct it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,250 @@
+
+        leaderLatch.addListener(this);
+        curatorFramework.getConnectionStateListenable().addListener(listener);
+        leaderLatch.start();
+        treeCache.start();
```
What would be the benefit? Unless there is something, I'd like to keep it as is.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -0,0 +1,432 @@
+    }
+
+    @Nonnull
+    private static ZooKeeperMultipleComponentLeaderElectionDriver startLeaderElectionDriver(
```
We can also rename it into `createLeaderElectionDriver` if this helps.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -0,0 +1,432 @@
+            new EachCallbackWrapper<>(zooKeeperExtension);
+
+    @Test
+    public void testElectionDriverGainsLeadership() throws Exception {
```
Yes. I can add the suffix if this makes things clearer.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -0,0 +1,432 @@
+                                    ZooKeeperLeaderRetrievalDriver.LeaderInformationClearancePolicy
+                                            .ON_LOST_CONNECTION));
+
+            final TestingListener leaderRetrievalListener = new TestingListener();
```
True. Since this is unrelated to my changes I will leave this for a future follow up. I fear that making this renamings will screw up the fixing of other changes.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -0,0 +1,432 @@
+        try {
+            leaderElectionListener.await(IsLeaderEvent.class);
+
+            final String componentId = "barfoo";
```
Will change it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,250 @@
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Write leader information {} for {}.", leaderInformation, componentId);
+        }
+        if (!leaderLatch.hasLeadership() || leaderInformation.isEmpty()) {
```
Yes, I think we should also write `LeaderInformation.empty()`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -0,0 +1,432 @@
+    }
+
+    @Test
+    public void testPublishLeaderInformation() throws Exception {
```
I'll add `testPublishEmptyLeaderInformation` and `testNonLeaderCannotPublishLeaderInformation`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -0,0 +1,432 @@
+
+            final LeaderInformation leaderInformation =
+                    LeaderInformation.known(UUID.randomUUID(), "foobar");
+            final String componentId = "barfoo";
```
Fair enough. I'll change it.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -0,0 +1,432 @@
+
+    @Test
+    public void testElectionDriverGainsLeadership() throws Exception {
+        final TestingLeaderElectionListener leaderElectionListener =
+                new TestingLeaderElectionListener();
+        final CuratorFrameworkWithUnhandledErrorListener curatorFramework = startCuratorFramework();
+
+        final ZooKeeperMultipleComponentLeaderElectionDriver leaderElectionDriver =
+                startLeaderElectionDriver(
+                        leaderElectionListener, curatorFramework.asCuratorFramework());
```
Let me check.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -0,0 +1,432 @@
+                        CompletableFuture.anyOf(
+                                electionDrivers.stream()
+                                        .map(ElectionDriver::getLeadershipFuture)
+                                        .collect(Collectors.toList())
+                                        .toArray(new CompletableFuture[0]));
```
Will change it.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/util/ZooKeeperUtilsITCase.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    @Nonnull
+    private CuratorFrameworkWithUnhandledErrorListener startCuratorFramework() {
```
I would leave this for a follow up task to introduce a `CuratorFrameworkExtension`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/leaderelection/LeaderElectionEvent.java:
```diff
@@ -0,0 +1,116 @@
+        return false;
+    }
+
+    public IsLeaderEvent asIsLeaderEvent() {
+        return as(IsLeaderEvent.class);
+    }
+
+    public NotLeaderEvent asNotLeaderEvent() {
+        return as(NotLeaderEvent.class);
+    }
+
+    public LeaderInformationChangeEvent asLeaderInformationChangeEvent() {
+        return as(LeaderInformationChangeEvent.class);
+    }
+
+    public AllKnownLeaderInformationEvent asAllKnownLeaderInformationEvent() {
+        return as(AllKnownLeaderInformationEvent.class);
+    }
```
True. I'll remove them.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/KubernetesUtils.java:
```diff
@@ -500,5 +502,49 @@ public static String tryToGetPrettyPrintYaml(KubernetesResource kubernetesResour
     }
 
+    private static final String LEADER_PREFIX = "org.apache.flink.k8s.leader.";
+    public static final char LEADER_INFORMATION_SEPARATOR = ',';
```
Yes, this is a good idea.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/KubernetesUtils.java:
```diff
@@ -500,5 +502,49 @@ public static String tryToGetPrettyPrintYaml(KubernetesResource kubernetesResour
         TASK_MANAGER
     }
 
+    private static final String LEADER_PREFIX = "org.apache.flink.k8s.leader.";
+    public static final char LEADER_INFORMATION_SEPARATOR = ',';
```
Let's stick to the existing conventions. I'll move it up.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/KubernetesUtils.java:
```diff
@@ -500,5 +502,49 @@ public static String tryToGetPrettyPrintYaml(KubernetesResource kubernetesResour
+    public static final char LEADER_INFORMATION_SEPARATOR = ',';
+
+    public static String encodeLeaderInformation(LeaderInformation leaderInformation) {
+        Preconditions.checkState(!leaderInformation.isEmpty());
```
True. I will change it.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/KubernetesUtils.java:
```diff
@@ -500,5 +502,49 @@ public static String tryToGetPrettyPrintYaml(KubernetesResource kubernetesResour
+                + leaderInformation.getLeaderAddress();
+    }
+
+    public static LeaderInformation parseLeaderInformationSafely(String value) {
```
I think this does not make a huge difference. But can do.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/utils/KubernetesUtils.java:
```diff
@@ -500,5 +502,49 @@ public static String tryToGetPrettyPrintYaml(KubernetesResource kubernetesResour
+        return LeaderInformation.known(leaderSessionId, leaderAddress);
+    }
+
+    @Nonnull
+    public static String createSingleLeaderKey(String componentId) {
+        return LEADER_PREFIX + componentId;
+    }
+
+    public static boolean isSingleLeaderKey(String key) {
+        return key.startsWith(LEADER_PREFIX);
+    }
+
+    @Nonnull
```
I don't think so. Will remove them.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/zookeeper/ZooKeeperMultipleComponentLeaderElectionHaServices.java:
```diff
@@ -0,0 +1,179 @@
+                multipleComponentLeaderElectionService =
+                        new DefaultMultipleComponentLeaderElectionService(
+                                fatalErrorHandler,
+                                "Single leader election service.",
```
Let me try to remove the leader description.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,268 @@
+
+    private final KubernetesSharedWatcher.Watch kubernetesWatch;
+
+    private AtomicBoolean running = new AtomicBoolean(true);
```
True. Will update it.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,268 @@
+                configMapSharedWatcher.watch(
+                        configMapName, new ConfigMapCallbackHandlerImpl(), watchExecutor);
+
+        leaderElector.run();
```
I don't see a strong benefit other than complicating the interface. Therefore, I would like to not do it at this point in time.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,268 @@
+            if (KubernetesUtils.isSingleLeaderKey(key)) {
+                final String leaderName = KubernetesUtils.extractLeaderName(key);
+                final LeaderInformation leaderInformation =
+                        KubernetesUtils.parseLeaderInformationSafely(keyValuePair.getValue());
```
I think this is ok because the corresponding `DefaultLeaderElectionService` would issue another publish command if it is the leader and its leader information is empty.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesMultipleComponentLeaderElectionHaServices.java:
```diff
@@ -0,0 +1,253 @@
+
+    @Override
+    protected String getLeaderPathForResourceManager() {
+        return "resourcemanager";
```
`KubernetesHaServices` will hopefully be removed in the next version. Hence, this will probably be work that can be spared.

tillrohrmann commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesHighAvailabilityRecoverFromSavepointITCase.java:
```diff
@@ -248,5 +229,24 @@ public void run(SourceContext<Integer> ctx) throws Exception {
+
+        @Override
+        public void initializeState(FunctionInitializationContext context) throws Exception {
+            final ListState<Integer> hasExecutedBeforeState =
+                    context.getOperatorStateStore()
+                            .getUnionListState(hasExecutedBeforeStateDescriptor);
+
+            // if we have state, then we resume from a savepoint --> stop the execution then
+            if (hasExecutedBeforeState.get().iterator().hasNext()) {
+                running = false;
+            }
+
+            hasExecutedBeforeState.clear();
+            // mark this subtask as executed before
+            hasExecutedBeforeState.add(getRuntimeContext().getIndexOfThisSubtask());
```
Yes, this is the idea of this state. Can rename it.

tillrohrmann commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesHighAvailabilityRecoverFromSavepointITCase.java:
```diff
@@ -248,5 +229,24 @@ public void run(SourceContext<Integer> ctx) throws Exception {
+                running = false;
+            }
+
+            hasExecutedBeforeState.clear();
```
It is to have a clean state into which we insert the subtask index.

tillrohrmann commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesHighAvailabilityRecoverFromSavepointITCase.java:
```diff
@@ -141,20 +130,8 @@ public void testRecoverFromSavepoint() throws Exception {
-                KubernetesCheckpointStoreUtil.INSTANCE.checkpointIDToName(2L);
-        assertThat(optional.get().getData().get(checkpointIdKey), is(notNullValue()));
-        assertThat(optional.get().getData().get(Constants.CHECKPOINT_COUNTER_KEY), is("3"));
+        assertThat(clusterClient.requestJobResult(jobId).join().isSuccess()).isTrue();
```
I'll try to do this once all the fixup commits are squashed.

tillrohrmann replied:
I've factored the changes out into a separate commit: fd57ecac15c1d0e8baa1059276178c84736d3558.

tillrohrmann commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesTestFixture.java:
```diff
@@ -0,0 +1,236 @@
+        return createFlinkKubeClientBuilder().build();
+    }
+
+    TestingFlinkKubeClient.Builder createFlinkKubeClientBuilder() {
```
I rather would like to not do this because I don't see an immediate benefit. If this should be necessary in the future, then let's spend the time on it.

tillrohrmann commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesLeaderElectionAndRetrievalITCase.java:
```diff
@@ -53,6 +51,9 @@
  */
 public class KubernetesLeaderElectionAndRetrievalITCase extends TestLogger {
```
I think this is a follow up task.

tillrohrmann commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -56,59 +62,179 @@
+            {
+                runTest(
+                        () -> {
+                            leaderCallbackGrantLeadership();
```
This is not a test but the actual signal to give leadership to the contender.

tillrohrmann commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -56,59 +62,179 @@
+            {
+                runTest(
+                        () -> {
+                            leaderCallbackGrantLeadership();
```
This is necessary to give leadership to the contender.

tillrohrmann commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesMultipleComponentLeaderElectionDriverTest.java:
```diff
@@ -56,59 +62,179 @@
+                            leaderElectionListener.await(LeaderElectionEvent.IsLeaderEvent.class);
+                            final LeaderInformation leaderInformation =
+                                    LeaderInformation.known(UUID.randomUUID(), "localhost");
+                            final String componentId = "foobar";
```
I'll change them.

tillrohrmann commented on flink-core/src/main/java/org/apache/flink/configuration/HighAvailabilityOptions.java:
```diff
@@ -204,6 +204,23 @@
+     * <p>Ideally, we can remove this option together with the old implementations in the next
+     * release.
+     */
+    @Documentation.Section(Documentation.Sections.EXPERT_HIGH_AVAILABILITY)
```
Yes. I think this is a good idea. I will create a ticket for this removal and add it in the comment.

tillrohrmann commented on flink-kubernetes/src/test/java/org/apache/flink/kubernetes/highavailability/KubernetesHighAvailabilityRecoverFromSavepointITCase.java:
```diff
@@ -104,15 +102,6 @@ public void setup() throws Exception {
 
-    @After
-    public void teardown() throws Exception {
-        flinkKubeClient
```
True. Will remove it.

tillrohrmann commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/highavailability/KubernetesCheckpointRecoveryFactory.java:
```diff
@@ -78,21 +84,46 @@ public CompletedCheckpointStore createRecoveredCompletedCheckpointStore(
-    public CheckpointIDCounter createCheckpointIDCounter(JobID jobID) {
-        return new KubernetesCheckpointIDCounter(
-                kubeClient, getConfigMapNameFunction.apply(jobID), lockIdentity);
+    public CheckpointIDCounter createCheckpointIDCounter(JobID jobID) throws Exception {
```
Then let's not do it ;-)

tillrohrmann commented:
Thanks for the review @XComp. I've resolved most of your comments. Those that I haven't resolved, I added an explanation for why not. Please take another look at this PR so that we can finish it swiftly :-)

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/util/ZooKeeperUtils.java:
```diff
@@ -402,6 +440,87 @@ public static ZooKeeperLeaderElectionDriverFactory createLeaderElectionDriverFac
+        final byte[] data;
+
+        if (leaderInformation.isEmpty()) {
+            data = null;
```
can we add a test case to `ZooKeeperUtilsTest` for writing data (including `null`) into the path? I did some digging into the Curator4 code tot check: It looks like it is. But it's hard to find information here. The [JavaDoc](https://www.javadoc.io/doc/org.apache.curator/curator-framework/4.0.0/org/apache/curator/framework/api/PathAndBytesable.html) is also not helpful. WDYT? Is it worth the effort?

tillrohrmann replied:
Do you mean a test where we write and read a `LeaderInformation.empty()`? This sounds like a good idea. If you meant a test where we use curator to write a `null` data, I think this not needed.

XComp replied:
Yup, agreed 👍

XComp reviewed (commented):
Thanks for addressing my comments, @tillrohrmann . I just have a few cosmetic comments but didn't approve it yet because of the `ZooKeeperUtilsTest` comment (in case you decide to add the test). In general, it looks good and I am ready to approve it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/DefaultMultipleComponentLeaderElectionService.java:
```diff
@@ -0,0 +1,278 @@
+                fatalErrorHandler,
+                multipleComponentLeaderElectionDriverFactory,
+                Executors.newSingleThreadExecutor(
+                        new ExecutorThreadFactory(String.format("leadershipOperationExecutor"))));
```
Good catch. Will remove it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,272 @@
+        }
+    }
+
+    private boolean shouldHandleLeaderInformationEvent(String removedNode) {
+        return running.get() && leaderLatch.hasLeadership() && isConnectionInfoNode(removedNode);
```
Good catch. Will change it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperMultipleComponentLeaderElectionDriver.java:
```diff
@@ -0,0 +1,272 @@
+        }
+    }
+
+    private String extractLeaderName(String path) {
```
I think this is logic specific to the `ZKMCLEDriver` because it leverages the fact that the path is always `/leader_name/connection_info`.

tillrohrmann commented:
Thanks for the review @XComp. I've addressed your comments. Moreover, I squashed all fixup commits and rebased onto the latest master. I hope that the PR is now ready to be merged.

XComp reviewed (approved):
LGTM 👍 The PR can be merged after the final CI run passed. Thanks for the contribution, @tillrohrmann . It was a pleasure to work with you! 🍾
